### ISSUE TITLE
Finding Qt6 in Cross-Compilation Setup

### ISSUE THREAD

**Issue by cordlandwehr** [mesonbuild/meson#13018]
**Describe the bug**
Currently testing with Meson 3.1.0 and Qt 6.7.0-rc1 (but I checked that no changes were done to the Qt support in meson in later versions); qt6 is detected via pkg-config method.

For a cross-compilation setup in Yocto there are typically two different sysroots (names are arbitrary, but I use the Yocto names):
- recipe-sysroot: contains all *.pc files, headers and cross-compiled libraries for the target machine
- recipe-sysroot-native: contains usr/bin/... and further folder structures that provide access to host tools to be used during the compilation

I am currently trying to e.g. build appstream with qt support or GStreamer with Qt support and both fail to find all Qt6 compile time dependencies (rcc, moc, etc.). The reason is very obvious when looking into the pc files provided by Qt:
```
prefix=/usr
exec_prefix=${prefix}
bindir=${prefix}/bin
libexecdir=${prefix}/libexec
libdir=${prefix}/lib
includedir=${prefix}/include

Name: Qt6 Core
Description: Qt Core module
Version: 6.7.0
Libs: -L${libdir} -lQt6Core
Cflags: -I${includedir}/QtCore -I${includedir} -DQT_CORE_LIB
Requires: Qt6Platform
```
AFAIS pkg-config has no way to know where to look for the host tools, because there is only the PKG_CONFIG_SYSROOT_DIR variable available, which points to the target sysroot.

**To Reproduce**
Currently, "easiest" is to compile appstream in Yocto master and activate the qt flag there.

**Expected behavior**
That the pkg-config method finds qt6

This change is related to #9791 , because specifically the libexec path makes the situation complicated. I am looking for feedback how to handle such an situation best to get the libexec and bin paths actually rooted into the host tools sysroot.

**Comment by dcbaker** [comment]
It sounds like the issue here is that you have two "sysroots" one for the build environment and one for the host environment?

**Comment by cordlandwehr** [comment]
Exactly, both for different architectures whereas one is the architecture of my host system on which I compile and the other is the architecture of my target embedded device.
For now I solved this by creating a meson-qt6.cross file 
```
cat > ${WORKDIR}/meson-qt6.cross <<EOF
[binaries]
qmake = '${STAGING_DIR_NATIVE}/usr/bin/qmake'
moc = '${STAGING_DIR_NATIVE}/usr/libexec/moc'
rcc = '${STAGING_DIR_NATIVE}/usr/libexec/rcc'
uic = '${STAGING_DIR_NATIVE}/usr/libexec/uic'
lrelease = '${STAGING_DIR_NATIVE}/usr/bin/lrelease'
EOF
```
and appending it as an additional cross file to the meson call via `--cross ${WORKDIR}/meson-qt6.cross`.
From my basic Meson knowledge, *think* that this is basically the right approach for my situation (but would be glad to have this confirmed).

Yet, what I think that is missing currently in the qt6 Meson logic is to automatically deduce all paths for moc, rcc, uic, lrelease from just the information where the qmake binary can be found. From the Qt perspective, qmake (alternatively, one could also use the qtpaths binary) provides the prefixing logic for the location of other host architecture binaries for cross-compilation setups.

**Comment by mschwan-phytec** [comment]
I have pretty much the same issue and worked around this by prepending `$PATH` with the native sysroot and libexec directory. E.g. when building a Qt6 application via a BitBake recipe:
```bitbake
# Qt tools like rcc, moc and uic are located in /usr/libexec, instead of
# /usr/bin, which is not in PATH by default.
do_configure:prepend:class-target() {
    export PATH=${STAGING_DIR_NATIVE}${libexecdir}:$PATH
}
```
Or when cross-compiling a Qt6 application with a Yocto-based SDK, I change `$PATH` before setting up the build:
```shell
export PATH=$OECORE_NATIVE_SYSROOT/usr/libexec:$PATH

# following meson setup and build
meson setup build
meson compile -C build
```
However, this does not feel right. If Qt6 only would expect the binaries to be in a directory included in `$PATH`, like they did in the beginning, these kind of issues wouldn't exist. I am unsure how and where to handle this properly.

**Comment by rossburton** [comment]
I've literally looked at `dependencies.qt.py` for thirty seconds, but isn't the problem here:

```
    def get_pkgconfig_host_libexecs(core: PkgConfigDependency) -> str:
        # Qt6 pkg-config for Qt defines libexecdir from 6.3+
        return core.get_variable(pkgconfig='libexecdir')
```

That logic assumes a non-cross build and at some point there needs to be, in effect, `native: true` to pivot to the correct sysroot/environment.

**Comment by mschwan-phytec** [comment]
@rossburton Thank you for your input!

Has anybody a rough idea how switching to native paths in `get_pkgconfig_host_libexecs()` and similar functions of `Qt6PkgConfigDependency` could be implemented? I had a look at the source code but couldn't quite figure out, how to implement this, as I am not familiar enough with it.

**Comment by newbluemoon** [comment]
Hi, I’ve ran into this issue (I think) trying to update sonic-visualiser on Void Linux (cf. https://github.com/void-linux/void-packages/pull/52406). When trying to cross-compile with different host and target word sizes, meson picks the target host tools (e.g. located under `.../lib32/...`) instead of the host’s (e.g. under `.../lib64/...`) via pkg-config.
There seems to be no clean solution how to handle cases like this. Would be nice to have an option to differentiate between host and target Qt6, or to simply set a custom host tools path.

**Comment by wellu** [comment]
Same problem over here. I'm trying to compile gstreamer1.0-plugins-good with Qt6 support in Yocto.

**Comment by wellu** [comment]
Update to Walnascar, same issue persists.
